webbpsf/opds.py

```diff
@@ -3207,8 +3206,76 @@ def coeffs_to_seg_state(coeffs):
     seg_state[:,1] = coeffs_tab[:,2]  # tilt in 2nd column
     return seg_state*1e6   # convert from meters & radians to micro units
 
+
+@functools.lru_cache
+def _get_lom(npix):
+    """Initialize a few things that are computationally expensive so we cache for multiple reuses"""
+
+    ote_lom = OTE_Linear_Model_WSS(npix=npix)
+    jw_ptt_basis = JWST_WAS_PTT_Basis()
+    return ote_lom, jw_ptt_basis
+
+
+def decompose_opd_segment_PTT(opd, plot=False):
+    """Phase decomposition of an OPD into PMSA piston, tip, tilt modes
+
+    Parameters
+    ----------
+    opd : 2d ndarray
+        OPD array
+    plot : bool
+        Display diagnostic plots in addition to doing the fit
+
+
+    Returns
+    -------
+
+    fit_opd : 2d ndarray
+        Projection of the input OPD into JWST PTT modes
+    coeffs : float array
+        Coefficients per mode, in order corresponding to the webbpsf.opds.JWST_WAS_PTT_Basis class,
+        which is {piston, xtilt, ytilt} repeated per segments in order
+
+    """
+    npix = opd.shape[0]
+
+    ote_lom, jw_ptt_basis = _get_lom(npix)
+
+    if ote_lom.opd.shape[0] != npix:
+        ote_lom = webbpsf.opds.OTE_Linear_Model_WSS(npix=npix)
+
+    combined_mask = ((ote_lom.amplitude != 0) & np.isfinite(opd))
+
+
+    coeffs = poppy.zernike.opd_expand_segments(opd, aperture=combined_mask,
+                                               basis=jw_ptt_basis, nterms=54,iterations=4)
+    fit = poppy.zernike.compose_opd_from_basis(coeffs, basis=jw_ptt_basis, npix=npix)
+
+    fit[~combined_mask]=np.nan
+    if plot:
+        fig, ax = plt.subplots(figsize=(16,4), nrows=1, ncols=1)
+        show_opd_image(np.hstack((opd, fit, opd-fit)),ax=ax )
```

Comment:
`show_opd_image` is not in OPDS, it needs to be initialized `webbpsf.trending.show_opd_image`.

Reply:
👍